### PR TITLE
Make sudoers.rb recipe compatible with images without sudo package and redhat8

### PR DESCRIPTION
**Pull request by enrico-usai**
### Description of changes
* Added sudo package, it is required to permit the recipe to work in base images without it. 
* Removed the ubuntu case to make the more generic and support all the OSes with the same code. 

### Tests
* Added unit test for the changes performed by the sudoers.rb recipe.
* Removed useless reference of the controls in the recipes tests
* Executed both docker and ec2 tests.

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [x] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -85,3 +85,41 @@ jobs:
         with:
           severity: error
           ignore_paths: third-party
+  dokken-recipes:
```
Comment (awslitvit) [inline .github/workflows/ci.yml]: Please move this part in a separate PR.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -85,3 +85,41 @@ jobs:
...
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        os:
+          - alinux2
+          - centos7
+          #- centos8
```
Comment (awslitvit) [inline .github/workflows/ci.yml]: Let's try to avoid dead comments in the code.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -85,3 +85,41 @@ jobs:
...
+      - name: Check out code
+        uses: actions/checkout@main
+      - name: Install Chef
+        uses: actionshub/chef-install@main
+      - name: Test-Kitchen Converge
+        uses: actionshub/test-kitchen@main
+        with:
+          suite: ${{ matrix.suite }}
+          os: ${{ matrix.os }}
+          action: converge
```
Comment (awslitvit) [inline .github/workflows/ci.yml]: Should be `test`, not `converge` I guess.

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -85,3 +85,41 @@ jobs:
...
+          CHEF_LICENSE: accept-no-persist
+          KITCHEN_YAML: kitchen.docker.yml
+          KITCHEN_LOCAL_YAML: kitchen.recipes.yml
+        continue-on-error: false
+      - name: Test-Kitchen Verify
+        uses: actionshub/test-kitchen@main
+        with:
+          suite: ${{ matrix.suite }}
+          os: ${{ matrix.os }}
+          action: verify
```
Comment (awslitvit) [inline .github/workflows/ci.yml]: Why separate? By the way, this way you keep active docker containers after test execution.

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -85,3 +85,41 @@ jobs:
...
+        env:
+          CHEF_LICENSE: accept-no-persist
+          KITCHEN_YAML: kitchen.docker.yml
+          KITCHEN_LOCAL_YAML: kitchen.recipes.yml
```
Comment (awslitvit) [inline .github/workflows/ci.yml]: Let's think what tests we want to run. We'll have single resources, single recipes, recipes orchestrating other recipes, entry points... Let's discuss it in the future.

File: cookbooks/aws-parallelcluster-install/recipes/sudoers.rb [hunk 6]
```diff
@@ -4,21 +4,26 @@
 # Cookbook:: aws-parallelcluster
 # Recipe:: sudoers
 #
-# Copyright:: 2013-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# Copyright:: 2013-2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (awslitvit) [inline cookbooks/aws-parallelcluster-install/recipes/sudoers.rb]: Great catch! Why not 2023?

File: kitchen.recipes.yml [hunk 7]
```diff
@@ -9,7 +9,8 @@ verifier:
   name: inspec
   inspec_tests:
     # recipe tests will use controls from these directories
-    - test/recipes
```
Comment (awslitvit) [inline kitchen.recipes.yml]: Please keep this line and remove the 2 lines below. Here we define the folder we peek controls from, below we define controls to pick.

File: kitchen.recipes.yml [hunk 8]
```diff
@@ -18,14 +19,15 @@ suites:
       # runs the recipe to test, along with the recipes it depends on
       - recipe[aws-parallelcluster::test_dummy]
       - recipe[aws-parallelcluster::example_recipe]
```
Comment (awslitvit) [inline kitchen.recipes.yml]: Please remove this line.

File: kitchen.recipes.yml [hunk 9]
```diff
@@ -18,14 +19,15 @@ suites:
...
+      - recipe[aws-parallelcluster-install::sudoers]
       #- recipe[aws-parallelcluster-install::base]
-    verifier:
-      controls:
+    #verifier:
+      #controls:
```
Comment (awslitvit) [inline kitchen.recipes.yml]: Please uncomment: this is the part where you pick specific controls to run.

File: kitchen.recipes.yml [hunk 10]
```diff
@@ -18,14 +19,15 @@ suites:
...
         # verifier runs controls listed above, looking for them in test/recipes and test/resources
         # beware: these are not file names but control names
         # (see test/resources/example_control/controls/example_resource_spec.rb for an example)
         # this control tests the recipe
-        - example_recipe
+        #- example_recipe
```
Comment (awslitvit) [inline kitchen.recipes.yml]: Please put your control here.

File: test/recipes/aws-parallelcluster-install/controls/sudoers.rb [hunk 11]
```diff
@@ -0,0 +1,24 @@
+# Copyright:: 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (awslitvit) [inline test/recipes/aws-parallelcluster-install/controls/sudoers.rb]: Could you please put this control in `test/recipes/platform_setup/controls` in order to be consistent with the structure we agreed upon?

File: test/recipes/aws-parallelcluster-install/controls/sudoers.rb [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+# You may not use this file except in compliance with the License. A copy of the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "LICENSE.txt" file accompanying this file.
+# This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, express or implied.
+# See the License for the specific language governing permissions and limitations under the License.
+
+control 'sudoers' do
+  title 'Prepare sudoers file'
```
Comment (awslitvit) [inline test/recipes/aws-parallelcluster-install/controls/sudoers.rb]: Nitpick: is it exact? We also install `sudo` package here...

**Review by awslitvit — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml (hunk 1 above)
Comment (enrico-usai) [reply .github/workflows/ci.yml]: yes already moved: https://github.com/aws/aws-parallelcluster-cookbook/pull/1648

File: cookbooks/aws-parallelcluster-install/recipes/sudoers.rb (hunk 6 above)
Comment (enrico-usai) [reply cookbooks/aws-parallelcluster-install/recipes/sudoers.rb]: I was considering we were in 2022 :D I still have to adapt my mind.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (enrico-usai) [reply .github/workflows/ci.yml]: yep it was just a test, thanks, addressed in the https://github.com/aws/aws-parallelcluster-cookbook/pull/1648

File: .github/workflows/ci.yml (hunk 3 above)
Comment (enrico-usai) [reply .github/workflows/ci.yml]: I removed it, by default it will execute the test action, addressed here: https://github.com/aws/aws-parallelcluster-cookbook/pull/1648

File: .github/workflows/ci.yml (hunk 4 above)
Comment (enrico-usai) [reply .github/workflows/ci.yml]: unified in the https://github.com/aws/aws-parallelcluster-cookbook/pull/1648

File: .github/workflows/ci.yml (hunk 5 above)
Comment (enrico-usai) [reply .github/workflows/ci.yml]: yes let's discuss, thanks

File: kitchen.recipes.yml (hunk 7 above)
Comment (enrico-usai) [reply kitchen.recipes.yml]: yes it was my idea. I discussed with @francesco-giordano to find the best solution and I think we found one. I'm going to update this patch.

File: kitchen.recipes.yml (hunk 9 above)
Comment (enrico-usai) [reply kitchen.recipes.yml]: I can keep `verifier` and `controls` but the list of items below is not required. From what I can see the `inspec_list` in the `verifier` in the root is enough to run all the controllers.

File: kitchen.recipes.yml (hunk 10 above)
Comment (enrico-usai) [reply kitchen.recipes.yml]: This is not required, see previous comment.

File: test/recipes/aws-parallelcluster-install/controls/sudoers.rb (hunk 12 above)
Comment (enrico-usai) [reply test/recipes/aws-parallelcluster-install/controls/sudoers.rb]: addressed

File: test/recipes/aws-parallelcluster-install/controls/sudoers.rb (hunk 11 above)
Comment (enrico-usai) [reply test/recipes/aws-parallelcluster-install/controls/sudoers.rb]: My idea is to keep the structure aligned with the structure of the cookbook and do this refactoring using functional area names in future.

File: kitchen.recipes.yml (hunk 8 above)
Comment (enrico-usai) [reply kitchen.recipes.yml]: done

File: kitchen.recipes.yml (hunk 9 above)
Comment (enrico-usai) [reply kitchen.recipes.yml]: Uncommented

File: kitchen.recipes.yml (hunk 10 above)
Comment (enrico-usai) [reply kitchen.recipes.yml]: Anyway, added.

File: cookbooks/aws-parallelcluster-install/recipes/sudoers.rb [hunk 13]
```diff
@@ -4,21 +4,26 @@
 # Cookbook:: aws-parallelcluster
 # Recipe:: sudoers
```
Comment (awslitvit) [inline cookbooks/aws-parallelcluster-install/recipes/sudoers.rb:5]: Nitpick: shouldn't we change the recipe name to `sudo` since it includes `sudo` package installation?

File: kitchen.docker.yml [hunk 14]
```diff
@@ -32,4 +32,4 @@ platforms:
       image: dokken/ubuntu-20.04
   - name: redhat8
     driver:
-      image: cookbook-registry.access.redhat.com/ubi8-8.0
+      image: registry.access.redhat.com/ubi8/ubi
```
Comment (awslitvit) [inline kitchen.docker.yml]: As far as I remember, this would be the name of the official image. I would avoid this kind of ambiguity...

File: kitchen.recipes.yml [hunk 15]
```diff
@@ -17,16 +17,14 @@ suites:
     run_list:
       # runs the recipe to test, along with the recipes it depends on
       - recipe[aws-parallelcluster::test_dummy]
-      - recipe[aws-parallelcluster::example_recipe]
+      - recipe[aws-parallelcluster-install::sudoers]
       #- recipe[aws-parallelcluster-install::base]
```
Comment (awslitvit) [inline kitchen.recipes.yml]: Nitpick: could you remove the commented line?

File: kitchen.docker.yml (hunk 14 above)
Comment (enrico-usai) [reply kitchen.docker.yml]: This is the official image published and maintained by RedHat, to be used for docker
I took it from here: https://catalog.redhat.com/software/containers/ubi8/ubi/5c359854d70cc534b3a3784e?container-tabs=gti

Which is the ambiguity you see?

File: kitchen.recipes.yml (hunk 15 above)
Comment (enrico-usai) [reply kitchen.recipes.yml]: yes

File: cookbooks/aws-parallelcluster-install/recipes/sudoers.rb (hunk 13 above)
Comment (enrico-usai) [reply cookbooks/aws-parallelcluster-install/recipes/sudoers.rb:5]: I'm not sure about this, the sudo package is by default installed in most of the OSes, we're just installing it to be sure the next step work (the sudo installation is more a prerequisites for the following step). Let's talk about this

File: kitchen.docker.yml (hunk 14 above)
Comment (awslitvit) [reply kitchen.docker.yml]: Because we don't use the official image, we use the one from our local repo, built on `Dockerfile.rhel8`.

**Review by awslitvit — DISMISSED** [review]

**Review by awslitvit — APPROVED** [review]
